src/t_hash.c

```diff
@@ -864,51 +841,157 @@ int hashTypeExists(redisDb *db, robj *o, sds field, int hfeFlags, int *isHashDel
+            if (hashTypeLength(o, 0) > server.hash_max_listpack_entries)
+                hashTypeConvert(o, OBJ_ENCODING_HT, &db->hexpires);
+        }
+
+    } else if (o->encoding == OBJ_ENCODING_HT) {
+        hfield newField = hfieldNew(field, sdslen(field), 0);
+        dict *ht = o->ptr;
+        dictEntry *de, *existing;
+
+        /* stored key is different than lookup key */
+        dictUseStoredKeyApi(ht, 1);
+        de = dictAddRaw(ht, newField, &existing);
+        dictUseStoredKeyApi(ht, 0);
+
+        /* If field already exists, then update "field". "Value" will be set afterward */
+        if (de == NULL) {
+            if (flags & HASH_SET_KEEP_FIELD) {
+                /* Not keep old field along with TTL */
+                hfieldFree(newField);
+            } else {
+                /* If attached TTL to the old field, then remove it from hash's private ebuckets */
+                hfield oldField = dictGetKey(existing);
+                hfieldPersist(o, oldField);
+                hfieldFree(oldField);
+                dictSetKey(ht, existing, newField);
+            }
+            sdsfree(dictGetVal(existing));
+            update = 1;
+            de = existing;
+        }
+
+        if (flags & HASH_SET_TAKE_VALUE) {
+            dictSetVal(ht, de, value);
+            flags &= ~HASH_SET_TAKE_VALUE;
+        } else {
+            dictSetVal(ht, de, sdsdup(value));
+        }
+    } else {
+        serverPanic("Unknown hash encoding");
+    }
+
+    /* Free SDS strings we did not referenced elsewhere if the flags
+     * want this function to be responsible. */
+    if (flags & HASH_SET_TAKE_FIELD && field) sdsfree(field);
+    if (flags & HASH_SET_TAKE_VALUE && value) sdsfree(value);
+    return update;
 }
 
-SetExRes hashTypeSetExpiry(HashTypeSetEx *ex, sds field, uint64_t expireAt, dictEntry **de) {
-    dict *ht = ex->hashObj->ptr;
+SetExRes hashTypeSetExpiryHT(HashTypeSetEx *exInfo, sds field, uint64_t expireAt) {
+    dict *ht = exInfo->hashObj->ptr;
     dictEntry *newEntry = NULL, *existingEntry = NULL;
 
     /* New field with expiration metadata */
     hfield hfNew = hfieldNew(field, sdslen(field), 1 /*withExpireMeta*/);
 
-    if ((ex->fieldSetCond == FIELD_DONT_CREATE) || (ex->fieldSetCond == FIELD_DONT_CREATE2)) {
-        if ((existingEntry = dictFind(ht, field)) == NULL) {
-            hfieldFree(hfNew);
-            return (ex->fieldSetCond == FIELD_DONT_CREATE) ?
-                   HSETEX_NO_CONDITION_MET : HSETEX_NO_FIELD;
-        }
-    } else {
-        dictUseStoredKeyApi(ht, 1);
-        newEntry = dictAddRaw(ht, hfNew, &existingEntry);
-        dictUseStoredKeyApi(ht, 0);
+    if ((existingEntry = dictFind(ht, field)) == NULL) {
+        hfieldFree(hfNew);
+        return HSETEX_NO_FIELD;
     }
 
     if (newEntry) {
-        *de = newEntry;
-
-        if (ex->expireSetCond & (HFE_XX | HFE_LT | HFE_GT)) {
+        if (exInfo->expireSetCond & (HFE_XX | HFE_LT | HFE_GT)) {
```

Comment:
`newEntry` is always NULL at this point right? You can discard that branch.
Secondly, this check should not involve HFE_LT if field does not exist, no?

Reply:
More dead, bad, code. Thanks!